---
title: "Fly"
author: "Zhang Yongchao"
date: "2025-04-15"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(rtracklayer)
library(devtools)
library(dplyr)
library(AnnotationDbi)
library(org.Dm.eg.db)
library(ZhangRtools)
display_venn <- function(x, ...){
  library(VennDiagram)
  grid.newpage()
  venn_object <- venn.diagram(x, filename = NULL, ...)
  grid.draw(venn_object)
}
```

# 核心数据库

**下载GTF**

url:


**预处理**

仅保留基因水平的信息

```{r}
gtf <- as.data.frame(rtracklayer::import("../inst/extdata/Flybase/Genome/dmel-all-r6.62.gtf.gz"))
table(gtf$type)
gtf <- gtf[gtf$type %in% c("gene"),c("gene_id", "gene_symbol", "type", "seqnames", "start", "end", "width", "strand" )]
head(gtf)
```



```{r}
colnames(gtf) <-  c("Ensembl", "Symbol","Gene.Type", "Chrosome", "Start", "End","Length", "Strand")
head(gtf)
```

注意： 基因名有重复值！**有些可能是因为在基因组上有多个拷贝**，或者是重复注释。

```{r}
table(is.na(gtf$Symbol)) # 基因名无缺失值
table(duplicated(gtf$Symbol)) # 没有重复的基因Symbol
head(sort(table(gtf$Symbol[duplicated(gtf$Symbol)]),decreasing = TRUE),20) # top 20 重复基因symbol,一般是lncRNA,asRNA等。
gtf.Duplicated.symbols <- unique(gtf$Symbol[duplicated(gtf$Symbol)])
# gtf.Duplicated.symbols
dplyr::arrange(gtf[gtf$Symbol %in% gtf.Duplicated.symbols,],desc(Symbol))
```

**注释全名、别名、蛋白名和Entrez编号**

注释不上的标记为NA值。

```{r}
gtf$EntrezID <- mapIds(x = org.Dm.eg.db, keys = gtf$Ensembl,column = "ENTREZID",keytype = "ENSEMBL",multiVals = "first")
gtf$Alias <- mapIds(x = org.Dm.eg.db, keys = gtf$Ensembl,column = "ALIAS",keytype = "ENSEMBL",multiVals = "first")
gtf$Name <- mapIds(x = org.Dm.eg.db, keys = gtf$Ensembl,column = "GENENAME",keytype = "ENSEMBL",multiVals = "first")
gtf$UniProt <- mapIds(x = org.Dm.eg.db, keys = gtf$Ensembl,column = "UNIPROT",keytype = "ENSEMBL",multiVals = "first")
gtf <- gtf[,c("Ensembl", "Symbol", "Gene.Type","EntrezID", "Name","Alias","UniProt","Chrosome", "Start", "End","Length", "Strand")]
```

```{r}
Flybase_Fly_Genes <- gtf
usethis::use_data(Flybase_Fly_Genes, overwrite = TRUE)
```
# 非核心数据库

## Flybase: Gene summary

https://ftp.flybase.net/releases/current/precomputed_files/genes/

```{r}
df <- read.delim("../inst/extdata/Flybase/Summary/best_gene_summary_fb_2025_01.tsv.gz",stringsAsFactors = FALSE)
head(df)
table(df$FBgn_ID %in% gtf$Ensembl)
Flybase_Fly_BestGeneSummary <- df
usethis::use_data(Flybase_Fly_BestGeneSummary, overwrite = TRUE)
```

```{r}
df <- as.data.frame(readr::read_tsv("../inst/extdata/Flybase/Summary/gene_snapshots_fb_2025_01.tsv.gz"))
head(df)
table(df$FBgn_ID %in% gtf$Ensembl)
Flybase_Fly_GeneSnapshots <- df
usethis::use_data(Flybase_Fly_GeneSnapshots, overwrite = TRUE)
```
```{r}
df <- as.data.frame(readr::read_tsv("../inst/extdata/Flybase/Summary/automated_gene_summaries.tsv.gz"))
head(df)
table(df$FBgn_ID %in% gtf$Ensembl)
Flybase_Fly_AutomatedGeneSummary <- df
usethis::use_data(Flybase_Fly_AutomatedGeneSummary, overwrite = TRUE)
```


## Animal TF DB v4.0

url: https://guolab.wchscu.cn/AnimalTFDB4/#/

**TFs**

url: https://guolab.wchscu.cn/AnimalTFDB4_static/download/TF_list_final/Drosophila_melanogaster_TF

```{r}
TF <- read.delim("../inst/extdata/AnimalTFDBv4/Drosophila_melanogaster/Drosophila_melanogaster_TF.txt",stringsAsFactors = FALSE)
head(TF)
TF <- TF[,c("Ensembl", "Entrez_ID", "Symbol" , "Family")]
```

根据EnsemblID、Symbol和EntrezID匹配基因名。

```{r}
TFs <- mapping_update(inputDF = TF, 
               db = gtf, 
               by.input = "Ensembl", by.db = "Ensembl", 
               by.input.2 = "Symbol", by.db.2 = "Symbol",
               by.input.3 = "Entrez_ID", by.db.3 = "EntrezID")
TFs$lost # 未匹配上的基因，经查，无法对应上。
```

```{r}
results <- TFs$matched
table(duplicated(results$Ensembl))
results$Symbol <- gtf$Symbol[match(results$Ensembl, gtf$Ensembl)]
results$Entrez <- gtf$EntrezID[match(results$Ensembl, gtf$Ensembl)]
colnames(results) <- c("Ensembl", "Entrez", "Symbol", "Family","EntrezID")
```

```{r}
AnimalTFDB_Fly_TF <- results
usethis::use_data(AnimalTFDB_Fly_TF, overwrite = TRUE)
```


**TF cofactor**

根据EnsemblID、Symbol和EntrezID匹配基因名。

```{r}
TF <- read.delim("../inst/extdata/AnimalTFDBv4/Drosophila_melanogaster/Drosophila_melanogaster_Cof.txt",stringsAsFactors = FALSE)
head(TF)
TF <- TF[,c("Ensembl", "Entrez_ID", "Symbol" , "Family")]
```

根据EnsemblID、Symbol和EntrezID匹配基因名。

```{r}
TFs <- mapping_update(inputDF = TF, 
               db = gtf, 
               by.input = "Ensembl", by.db = "Ensembl", 
               by.input.2 = "Symbol", by.db.2 = "Symbol",
               by.input.3 = "Entrez_ID", by.db.3 = "EntrezID")
TFs$lost # 未匹配上的基因，经查，无法对应上。
```


```{r}
results <- TFs$matched
table(results$label)
results$label <- NULL
results$Symbol <- gtf$Symbol[match(results$Ensembl, gtf$Ensembl)]
results$Entrez <- gtf$EntrezID[match(results$Ensembl, gtf$Ensembl)]
colnames(results) <- c("Ensembl", "Entrez", "Symbol", "Family","EntrezID")
```

```{r}
AnimalTFDB_Fly_TFCofactors <- results
usethis::use_data(AnimalTFDB_Fly_TFCofactors, overwrite = TRUE)
```

## GENEONTOLOGY

> [详解如何获取物种所有基因对应的GO注释](https://cloud.tencent.com/developer/article/1625243)

下载位置： https://current.geneontology.org/products/pages/downloads.html

https://geneontology.org/docs/download-ontology/ 页面下载： https://purl.obolibrary.org/obo/go/go-basic.obo

```{r}
df <- read.delim("../inst/extdata/GENEONTOLOGY/fb.gaf.gz",skip = 36,header = FALSE)
# https://geneontology.org/docs/go-annotation-file-gaf-format-2.2/
colnames(df) <- c("DB", "ProteinID", "Symbol","Qualifier","GOID","Reference", "EvidenceCode","WithOrFrom","Aspect", 
                  "GeneProducts","Synonym","Type","Taxon", "Date","AssignedBy","AnnotationExtension","GeneProductsFromID")
write.csv(unique(df$ProteinID), file = "fly.IDs.unique.csv",row.names = FALSE,quote = FALSE)
ID.map <- read.delim("./FlyBase_Fields_download_for_GENEONTOLOGY.txt",stringsAsFactors = FALSE)
lost <- ID.map$X.SUBMITTED.ID[ID.map$FBID_KEY == ""]
df <- df[!df$ProteinID %in% lost,]
ID.map <- ID.map[ID.map$FBID_KEY != "",]
mapping <- ID.map$FBID_KEY
names(mapping) <- ID.map$X.SUBMITTED.ID
df$ProteinID <- mapping[df$ProteinID]

library(ontologyIndex)
gs <- get_OBO("../inst/extdata/GENEONTOLOGY/go-basic.obo")
table(df$GOID %in% names(gs$name))
df$GOName <- gs$name[df$GOID]
```

```{r}
# 为什么不用sumbol，因为有30多个NA值
df.sub <- df[,c("ProteinID","GOID","GOName","Aspect")]
# remove 3 root: GO:0005575:cellular_component, GO:0003674:molecular_function, GO:0008150: biological_process
df.sub <- df.sub[!df.sub$GOID %in% c("GO:0005575", "GO:0003674", "GO:0008150"),]
res <- list()
aspect.name <- c("MolecularFunction","BiologicalProcess","CellularComponent")
names(aspect.name) <- c("F","P","C")
for (i in unique(df.sub$Aspect)) {
  df.sub.sub <- df.sub[df.sub$Aspect == i,]
  df.sub.sub.aggr <- Aggregate_df(df = df.sub.sub, id = "ProteinID")
  df.sub.sub.aggr$Aspect <- NULL
  colnames(df.sub.sub.aggr)[2:3] <- paste(aspect.name[i],colnames(df.sub.sub.aggr)[2:3],sep = "_")
  res[[aspect.name[i]]] <- df.sub.sub.aggr
}
```


```{r}
# saveRDS(res,file = "temp.rds")
# run codes bellow under a server!
# results <- merge(res$MolecularFunction, res$CellularComponent, all= TRUE, by = "Symbol")
# results <- merge(results, res$BiologicalProcess, all= TRUE, by = "Symbol")
# results[is.na(results)] <- "-"
```

```{r}
GENEONTOLOGY_Fly <- readRDS("./fly.go.rds")
map.df <- distinct(ID.map,FBID_KEY,SYMBOL)
mapping <- map.df$SYMBOL
names(mapping) <- map.df$FBID_KEY

GENEONTOLOGY_Fly$Symbol <- mapping[GENEONTOLOGY_Fly$ProteinID]
colnames(GENEONTOLOGY_Fly)[1] <- "Ensembl"

GENEONTOLOGY_Fly <- GENEONTOLOGY_Fly[,c( "Ensembl", "Symbol", "MolecularFunction_GOID",  "MolecularFunction_GOName", "CellularComponent_GOID", "CellularComponent_GOName", "BiologicalProcess_GOID" , "BiologicalProcess_GOName")]

load("../data/Flybase_Fly_Genes.rda")
table(GENEONTOLOGY_Fly$Symbol %in% Flybase_Fly_Genes$Symbol)
usethis::use_data(GENEONTOLOGY_Fly, overwrite = TRUE)
```
