---
title: "Mouse Data Preparation"
author: "Zhang Yongchao"
date: "2025-04-15"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

> Attention: 可以使用在线工具https://www.informatics.jax.org/batch进行小鼠基因名（支持混合类型）的更新和转换！
另外，发现很多MGI id不存在对应的Ensembl ID。

```{r}
library(devtools)
library(rtracklayer)
library(dplyr)
library(AnnotationDbi)
library(org.Mm.eg.db)
library(ZhangRtools)
display_venn <- function(x, ...){
  library(VennDiagram)
  grid.newpage()
  venn_object <- venn.diagram(x, filename = NULL, ...)
  grid.draw(venn_object)
}
```

# 核心数据库

**下载GTF**

url: https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/latest_release/

下载 gencode.vM34.primary_assembly.annotation.gtf.gz, 保存到 inst/extdata/GENCODE目录下。

**预处理**
 
```{r}
gtf <- as.data.frame(rtracklayer::import("../inst/extdata/GENCODE/gencode.vM36.annotation.gtf.gz"))
table(gtf$type)
gtf <- gtf[gtf$type %in% "gene",c("gene_id", "gene_name", "mgi_id", "gene_type", "seqnames", "start", "end", "width", "strand" )]
head(gtf)
```

```{r}
colnames(gtf) <-  c("Ensembl", "Symbol","MGI", "Gene.Type", "Chrosome", "Start", "End","Length", "Strand")
gtf$Ensembl <- gsub("\\.\\d+","",gtf$Ensembl)
gtf$MGI <- gsub("MGI:","",gtf$MGI)
head(gtf)
```

注意： 基因名有重复值！**有些可能是因为在基因组上有多个拷贝**，或者是重复注释。

```{r}
table(is.na(gtf$Symbol)) # 基因名无缺失值
table(duplicated(gtf$Symbol)) # 有1624个重复的基因Symbol
head(sort(table(gtf$Symbol[duplicated(gtf$Symbol)]),decreasing = TRUE),20) # top 20 重复基因symbol
gtf.Duplicated.symbols <- unique(gtf$Symbol[duplicated(gtf$Symbol)])
gtf.Duplicated.symbols
dplyr::arrange(gtf[gtf$Symbol %in% gtf.Duplicated.symbols,],desc(Symbol))
```

**注释全名、别名、蛋白名和Entrez编号**

注释不上的标记为NA值。

```{r}
gtf$EntrezID <- mapIds(x = org.Mm.eg.db, keys = gtf$Ensembl,column = "ENTREZID",keytype = "ENSEMBL",multiVals = "first")
gtf$Alias <- mapIds(x = org.Mm.eg.db, keys = gtf$Ensembl,column = "ALIAS",keytype = "ENSEMBL",multiVals = "first")
gtf$Name <- mapIds(x = org.Mm.eg.db, keys = gtf$Ensembl,column = "GENENAME",keytype = "ENSEMBL",multiVals = "first")
gtf$UniProt <- mapIds(x = org.Mm.eg.db, keys = gtf$Ensembl,column = "UNIPROT",keytype = "ENSEMBL",multiVals = "first")
gtf <- gtf[,c("Ensembl", "Symbol", "MGI", "Gene.Type","EntrezID", "Name","Alias","UniProt","Chrosome", "Start", "End","Length", "Strand")]
```

```{r}
GENCODE_Mouse_Genes <- gtf
usethis::use_data(GENCODE_Mouse_Genes, overwrite = TRUE)
```


## 非核心数据库

```{r}
load("../data/GENCODE_Mouse_Genes.rda")
```


## Animal TF DB v4.0

url: https://guolab.wchscu.cn/AnimalTFDB4/#/


**下载数据**

手动下载链接： https://guolab.wchscu.cn/AnimalTFDB4/#/Download

右键 - 链接另存为

**TFs**

```{r}
TF <- read.delim("../inst/extdata/AnimalTFDBv4/Mus_musculus/Mus_musculus_TF.txt",stringsAsFactors = FALSE)
head(TF)
TF <- TF[,c("Ensembl", "Entrez_ID", "Symbol" , "Family")]
```

根据EnsemblID、Symbol和EntrezID匹配基因名。

```{r}
# 注意TF里重复值symbol
arrange(TF[TF$Symbol %in% TF$Symbol[duplicated(TF$Symbol)],],Symbol)

TFs <- mapping_update(inputDF = TF, 
               db = gtf, 
               by.input = "Ensembl", by.db = "Ensembl", 
               by.input.2 = "Symbol", by.db.2 = "Symbol",
               by.input.3 = "Entrez_ID", by.db.3 = "EntrezID")
TFs$lost # 有20未匹配上的基因，经查，无法对应上。可能是ensembl新版的删除这些条目。
```


```{r}
results <- TFs$matched
table(results$label)
results$label <- NULL
results$Symbol <- gtf$Symbol[match(results$Ensembl, gtf$Ensembl)]
results$Entrez <- gtf$EntrezID[match(results$Ensembl, gtf$Ensembl)]

AnimalTFDB_Mouse_TF <- results
usethis::use_data(AnimalTFDB_Mouse_TF, overwrite = TRUE)
```

**TF cofactor**

根据EnsemblID、Symbol和EntrezID匹配基因名。

```{r}
TF <- read.delim("../inst/extdata/AnimalTFDBv4/Mus_musculus/Mus_musculus_Cof.txt",stringsAsFactors = FALSE)
head(TF)
TF <- TF[,c("Ensembl", "Entrez_ID", "Symbol" , "Family")]
```

根据EnsemblID、Symbol和EntrezID匹配基因名。

```{r}
TFs <- mapping_update(inputDF = TF, 
               db = gtf, 
               by.input = "Ensembl", by.db = "Ensembl", 
               by.input.2 = "Symbol", by.db.2 = "Symbol",
               by.input.3 = "Entrez_ID", by.db.3 = "EntrezID")
TFs$lost # 未匹配上的基因，经查，无法对应上。
```


```{r}
results <- TFs$matched
table(results$label)
results$label <- NULL
results$Symbol <- gtf$Symbol[match(results$Ensembl, gtf$Ensembl)]
results$Entrez <- gtf$EntrezID[match(results$Ensembl, gtf$Ensembl)]
# colnames(results) <- c("Ensembl", "Entrez", "Symbol", "Family")

AnimalTFDB_Mouse_TFCofactors <- results
usethis::use_data(AnimalTFDB_Mouse_TFCofactors, overwrite = TRUE)
```

## The IUPHAR/BPS Guide to PHARMACOLOGY

url: https://www.guidetopharmacology.org/index.jsp

Current Release Version 2022.2 (9th June 2022)

full datbase(dmp format, created with PostgreSQL version 9.2): https://www.guidetopharmacology.org/DATA/public_iuphardb_v2022.2.zip

or Read a description of the files: "https://www.guidetopharmacology.org/DATA/file_descriptions.txt"

Download complete **target and family** list: https://www.guidetopharmacology.org/DATA/targets_and_families.csv
Download complete **ligand** list: https://www.guidetopharmacology.org/DATA/ligands.csv
Download **ligand ID mapping** file: https://www.guidetopharmacology.org/DATA/ligand_id_mapping.csv
Download SDF file of **ligands** with SMILES: https://www.guidetopharmacology.org/DATA/all_ligands.sdf
Download detailed **endogenous/natural ligand** list: https://www.guidetopharmacology.org/DATA/detailed_endogenous_ligands.csv
Download condensed **endogenous/natural ligand** list: https://www.guidetopharmacology.org/DATA/endogenous_ligands.csv
Download **approved drugs with primary targets** list: https://www.guidetopharmacology.org/DATA/approved_drug_primary_target_interactions.csv
Download complete **peptide ligand** list: https://www.guidetopharmacology.org/DATA/peptides.csv
Download all interaction data for ligands and targets:https://www.guidetopharmacology.org/DATA/interactions.csv
Download a file mapping Guide to PHARMACOLOGY target and peptide ligand IDs and URLs to HGNC gene IDs and symbols: https://www.guidetopharmacology.org/DATA/GtP_to_HGNC_mapping.csv
Download a file mapping Guide to PHARMACOLOGY target IDs and URLs to UniProt accessions: https://www.guidetopharmacology.org/DATA/GtP_to_UniProt_mapping.csv



**target.and.family 中有哪些targets**

```{r}
target.and.family <- read.csv("../inst/extdata/GuideToPharmacology/targets_and_families.csv", stringsAsFactors = FALSE, skip = 1)
table(target.and.family$Type)
target.and.family_targets <- target.and.family[,c("Type","Family.name", "Target.id","MGI.id","MGI.symbol","Mouse.Entrez.Gene", "Mouse.Ensembl.Gene")]
target.and.family_targets$MGI.id <- gsub("^MGI:","", target.and.family_targets$MGI.id)

target.and.family_targets$MGI.id[is.na(target.and.family_targets$MGI.id)] <- ""
target.and.family_targets$MGI.symbol[is.na(target.and.family_targets$MGI.symbol)] <- ""
target.and.family_targets$Mouse.Entrez.Gene[is.na(target.and.family_targets$Mouse.Entrez.Gene)] <- ""
target.and.family_targets$Mouse.Ensembl.Gene[is.na(target.and.family_targets$Mouse.Ensembl.Gene)] <- ""

janitor::get_dupes(target.and.family_targets, Target.id) # 重复的Target ID，对应多个不同的family name，即ID重复的，Family.name是不一样的，别的都一样

# 第一步合并相同Target.id的Family.name
target.and.family_targets %>%
  dplyr::group_by(Target.id) %>%
  mutate(Family.name = paste0(unique(Family.name), collapse = ";"))  %>% # 修改Family.name列
  filter(row_number() == 1) %>%  # 仅保留第一行
  ungroup() -> target.and.family_targets

# 去除id，symbol，Entrez.Gene，Ensembl.Gene均为空的条目
target.and.family_targets <-  dplyr::filter(target.and.family_targets, trimws(MGI.id) != "" | trimws(MGI.symbol) != "" | trimws(Mouse.Entrez.Gene) != "" | trimws(Mouse.Ensembl.Gene) != "")


# MGI.id 有没有重复的？
table(is.na(target.and.family_targets$MGI.id))
table(target.and.family_targets$MGI.id %in% gtf$MGI)
table(target.and.family_targets$MGI.symbol %in% gtf$Symbol)
table(target.and.family_targets$Mouse.Entrez.Gene %in% gtf$EntrezID)
table(target.and.family_targets$Mouse.Ensembl.Gene %in% gtf$Ensembl)

# 目测用MGI ID匹配是最好的！
# 发现有些Ensembl ID重复项的MGI id并不同，经查，认定是 Ensembl ID 标记有错误。并且发现列表中有些基因罗列了多个Ensembl ID，因此考虑放弃使用Ensembl ID。**考虑使用MGI ID 和 Symbol做合并**！

target.and.family_targets.lost <- target.and.family_targets[!target.and.family_targets$MGI.id %in% gtf$MGI,]
target.and.family_targets.lost

# insert MGI id for empty values by Entrez ID
target.and.family_targets$MGI.id[target.and.family_targets$MGI.id == ""] <- gtf$MGI[match(target.and.family_targets$Mouse.Entrez.Gene[target.and.family_targets$MGI.id == ""], gtf$EntrezID)]
# expand the dataframe for some MGI id has multiple IDs sepearted by |
target.and.family_targets <- Expand_df(target.and.family_targets, id = 'MGI.id', splitby = "[|]")
target.and.family_targets$MGI.id <- gsub("^MGI:", "", target.and.family_targets$MGI.id)

table(target.and.family_targets$MGI.id %in% gtf$MGI)

target.and.family_targets.lost <-  target.and.family_targets[!target.and.family_targets$MGI.id %in% gtf$MGI,]
target.and.family_targets.lost$MGI.symbol %in% gtf$Symbol
target.and.family_targets.lost$Mouse.Entrez.Gene %in% gtf$EntrezID
target.and.family_targets$MGI.id[(!target.and.family_targets$MGI.id %in% gtf$MGI) & (target.and.family_targets$Mouse.Entrez.Gene %in% gtf$EntrezID)] <- gtf$MGI[match(target.and.family_targets$Mouse.Entrez.Gene[(!target.and.family_targets$MGI.id %in% gtf$MGI) & (target.and.family_targets$Mouse.Entrez.Gene %in% gtf$EntrezID)], gtf$EntrezID)]


targets.gene.unique <- target.and.family_targets[target.and.family_targets$MGI.id %in% gtf$MGI,]
all(targets.gene.unique$MGI.id %in% gtf$MGI)
table(duplicated(targets.gene.unique$MGI.id))


# 重复的MGI ID
janitor::get_dupes(targets.gene.unique, MGI.id) # remove the second
targets.gene.unique <- targets.gene.unique[!duplicated(targets.gene.unique$MGI.id),]

# 重复的target ID
janitor::get_dupes(targets.gene.unique, Target.id) # 重复的Target ID，对应多个不同的family name，即ID重复的，Family.name是不一样的，别的都一样
# 最终得到的target，有一个target id 可能对应多个MGI id

# gtf里的MGI id有重复值
gtf.Duplicated.MGI <- as.vector(na.omit(gtf$MGI[duplicated(gtf$MGI)]))
map.unique <- targets.gene.unique[!targets.gene.unique$MGI.id %in% gtf.Duplicated.MGI,]
map.duplicates <- targets.gene.unique[targets.gene.unique$MGI.id %in% gtf.Duplicated.MGI,]

gtf[gtf$MGI == "2442042" & !is.na(gtf$MGI),] # 1个MGI id对应两个Ensembl ID
targets.gene.unique[targets.gene.unique$MGI.id == "2442042",]

targets.gene.unique$MGI.symbol <- gtf$Symbol[match(targets.gene.unique$MGI.id, gtf$MGI)]
targets.gene.unique$Mouse.Entrez.Gene <- gtf$EntrezID[match(targets.gene.unique$MGI.id, gtf$MGI)]
targets.gene.unique$Mouse.Ensembl.Gene <- gtf$Ensembl[match(targets.gene.unique$MGI.id, gtf$MGI)]
new.line <- targets.gene.unique[targets.gene.unique$MGI.id == "2442042",]
new.line$Mouse.Ensembl.Gene <- "ENSMUSG00000115067"
targets.gene.unique <- rbind(targets.gene.unique, new.line)


table(duplicated(targets.gene.unique$Mouse.Ensembl.Gene)) # 均无重复值
table(duplicated(targets.gene.unique$MGI.id)) # 有一个重复值

targets.gene.unique <- targets.gene.unique[,c("Mouse.Ensembl.Gene", "MGI.id", "MGI.symbol", "Mouse.Entrez.Gene", "Target.id","Type", "Family.name")]
colnames(targets.gene.unique) <- c("Ensembl","MGI", "Symbol", "Entrez", "Target.ID","Type", "Family")

Mouse_GuideToPHARMACOLOGY_Targets <- targets.gene.unique
usethis::use_data(Mouse_GuideToPHARMACOLOGY_Targets, overwrite = TRUE)
```

```{r}
janitor::get_dupes(targets.gene.unique, Target.ID)
# 1. 同一Target ID对应多个不同的同系物
# 2. 同一Target ID对应多个不同的Eesembl ID，有相同的MGI ID和Symbol，基因组多拷贝？还是遗留的注释信息？
```


**ligand文件 中有哪些 ligand**

ligand.ID.mapping 与 ligand 两个文档内容是一致的； 注意：一个ligand ID可以对应多个基因; 一个基因ID也可以对应多个Ligand ID。

```{r}
ligand <- read.csv("../inst/extdata/GuideToPharmacology/ligands.csv", stringsAsFactors = FALSE, skip = 1)
ligand.ID.mapping <- read.csv("../inst/extdata/GuideToPharmacology/ligand_id_mapping.csv", stringsAsFactors = FALSE, skip = 1)

all(ligand.ID.mapping$Ensembl.ID == ligand$Ensembl.ID)
# ligand.ID.mapping 与ligand 里的Ensembl.ID是一样的，并且维度也一样。

# 提取所有鼠里的Ligands基因。
table(ligand.ID.mapping$Type)
table(duplicated(ligand.ID.mapping$Ligand.id)) # Liand id是没有重复值的！

ligands.id.uniqe <- ligand.ID.mapping[grepl("ENSMUSG", ligand.ID.mapping$Ensembl.ID),c("Ligand.id","Name", "Species" ,"Type","Ensembl.ID")]
table(ligands.id.uniqe$Type)

# 一个Ligand id 对应多个Ensembl id，并且去除非人基因
tmp <- c()
for (i in ligands.id.uniqe$Ensembl.ID) {
  if(grepl("|", i,fixed = TRUE)){
    tmp <- append(tmp, paste0(grep("^ENSMUSG",unlist(strsplit(i,split = "|",fixed = TRUE)),value = TRUE),collapse = ";"))
  }else{
    tmp <- append(tmp, i)
  }
}
ligands.id.uniqe$Ensembl.ID <- tmp

# 修正Ensembl IDs
## 制备old - new name数据库
ensembl.ids.unique <- unique(unlist(lapply(ligands.id.uniqe$Ensembl.ID, function(x)unlist(strsplit(x, split = ";",fixed = TRUE))))) # 提取所有IDs
ensembl.ids.unique.lost <- ensembl.ids.unique[!ensembl.ids.unique %in% gtf$Ensembl] # 提取不能被识别的

ligands.id.uniqe[ligands.id.uniqe$Ensembl.ID %in% ensembl.ids.unique.lost,]
ligands.id.uniqe[grepl("ENSMUSG00000055951", ligands.id.uniqe$Ensembl.ID),]
# corticotrophin-releasing hormone Human对应的 ENSMUSG0000004979 应该改为ENSMUSG00000049796
# CCL25 对应两个ENSMUSG00000023235;ENSMUSG00000055951，去除ENSMUSG00000055951就可以了。
ligands.id.uniqe$Ensembl.ID[grepl("ENSMUSG00000055951", ligands.id.uniqe$Ensembl.ID)] <- "ENSMUSG00000023235"
ligands.id.uniqe$Ensembl.ID[ligands.id.uniqe$Ensembl.ID %in% ensembl.ids.unique.lost] <- "ENSMUSG00000049796"

# 依据Esembl ID拆分和整理数据
janitor::get_dupes(ligands.id.uniqe, Ensembl.ID) # 发现有重复的Ensembl IDs，一个Ensembl ID对应多个ligand ID
# 按照Ensembl ID，将相同ligand ID的条目合并
ligands.id.uniqe %>%
  dplyr::group_by(Ensembl.ID) %>%
  mutate(Ligand.id = paste0(Ligand.id, collapse = ";"), # 合并id列
          Name = paste0(Name, collapse = ";"))  %>% # 合并name列
  filter(row_number()==1) %>%  # 仅保留第一行
  ungroup() -> ligands.gene.uniqe
table(ligands.gene.uniqe$Ensembl.ID %in% gtf$Ensembl)
ligands.gene.uniqe$Symbol <- gtf$Symbol[match(ligands.gene.uniqe$Ensembl.ID, gtf$Ensembl)]
ligands.gene.uniqe <- ligands.gene.uniqe[,c("Ensembl.ID", "Symbol","Ligand.id", "Name", "Type" )]
dim(ligands.gene.uniqe)
table(ligands.gene.uniqe$Type)
ligands.gene.uniqe$Name <- NULL
colnames(ligands.gene.uniqe) <- c("Ensembl","Symbol", "Ligand.ID","Type")

Mouse_GuideToPHARMACOLOGY_LigandsPeptides <- ligands.gene.uniqe
usethis::use_data
```


**detailed.endogenous.ligand** 注意有些行的ligand ID是空的

发现这里的endogenous ligand与前面的peptides ligand是一样的，这里不再重复输出。

```{r}
# detailed.endogenous.ligand <-  read.csv("../inst/extdata/GuideToPharmacology/detailed_endogenous_ligands.csv.gz", stringsAsFactors = FALSE, skip = 1)
# head(detailed.endogenous.ligand)
# detailed.endogenous.ligand[!detailed.endogenous.ligand$Ligand.ID %in% ligand$Ligand.ID,]
# detailed.endogenous.ligand.Mouse <- detailed.endogenous.ligand[detailed.endogenous.ligand$Interaction.Species == "Mouse",]
# # 提取所有唯一的ENSMUSG ID
# uniqe.ensembl.ids <- unique(trimws(unlist(lapply(grep("ENSMUSG",detailed.endogenous.ligand.Mouse$Ligand.Ensembl.Gene.ID,fixed = TRUE,value = TRUE),function(x)unlist(strsplit(x,split = "|",fixed = TRUE))))))
# length(uniqe.ensembl.ids) # 粗略估计这里的peptide ligand数目
# table(uniqe.ensembl.ids %in% ligands.gene.uniqe$Ensembl)
# table(uniqe.ensembl.ids %in% db.ligands.ensembl$old)
# all(detailed.endogenous.ligand.Mouse$Ligand.ID %in% all.ligands.and.targets.interactions$Ligand.ID) # 也都被包括到interactions中
```

**condensed.endogenous.ligand** 文档targets基本都是人的基因。

```{r}
# condensed.endogenous.ligand <- read.csv("../inst/extdata/GuideToPharmacology/detailed_endogenous_ligands.csv.gz", stringsAsFactors = FALSE, skip = 1)
# condensed.endogenous.ligand <- condensed.endogenous.ligand
# condensed.endogenous.ligand[!condensed.endogenous.ligand$Ligand.ID %in% ligands.id.uniqe$Ligand.ID,]
# # 一个Ligand id 对应多个Ensembl id，并且去除非人基因
# tmp <- c()
# for (i in condensed.endogenous.ligand$Ligand.Ensembl.Gene.ID) {
#   if(grepl("|", i,fixed = TRUE)){
#     tmp <- append(tmp, paste0(grep("^ENSMUSG",unlist(strsplit(i,split = "|",fixed = TRUE)),value = TRUE),collapse = ";"))
#   }else{
#     tmp <- append(tmp, i)
#   }
# }
# condensed.endogenous.ligand$Ligand.Ensembl.Gene.ID <- tmp
# tmp <- c()
# for (i in condensed.endogenous.ligand$Target.Ensembl.Gene.ID) {
#   if(grepl("|", i,fixed = TRUE)){
#     tmp <- append(tmp, paste0(grep("^ENSMUSG",unlist(strsplit(i,split = "|",fixed = TRUE)),value = TRUE),collapse = ";"))
#   }else{
#     tmp <- append(tmp, i)
#   }
# }
# condensed.endogenous.ligand$Target.Ensembl.Gene.ID <- tmp
# condensed.endogenous.ligand$Ligand.Symbol <- update_IDs(old = condensed.endogenous.ligand$Ligand.Ensembl.Gene.ID, db = gtf, from = "Ensembl", to = "Symbol", fixed = TRUE, split = ";")
# condensed.endogenous.ligand$Target.Symbol <- update_IDs(old = condensed.endogenous.ligand$Target.Ensembl.Gene.ID, db = gtf, from = "Ensembl", to = "Symbol", fixed = TRUE, split = ";")
# head(sort(table(condensed.endogenous.ligand$Ligand.Symbol),decreasing = TRUE),30)
# head(sort(table(condensed.endogenous.ligand$Target.Symbol),decreasing = TRUE),30)
# 
# # 不是所有的ligand基因都记录了targets
# table(ligands.id.uniqe$Ligand.id %in% condensed.endogenous.ligand$Ligand.ID)
```


**all.ligands.and.targets.interactions**

target_ligand：注意，这个表示这个target本来也是一个ligand，但是在这个条目中，他是一个target。

```{r}
all.ligands.and.targets.interactions <- read.csv("../inst/extdata/GuideToPharmacology/interactions.csv", stringsAsFactors = FALSE, skip = 1)
complete.peptide.ligand <-  read.csv("../inst/extdata/GuideToPharmacology/peptides.csv", stringsAsFactors = FALSE, skip = 1)
# 只保留鼠的
all.interactions.targets <- all.ligands.and.targets.interactions[all.ligands.and.targets.interactions$Target.Species == "Mouse",]
# 配体与配体的互作？这里不再进行深入探索，以后再说吧！  Target.Ligand 开头的列是啥？
target.id.NA <- dplyr::distinct(all.interactions.targets[is.na(all.interactions.targets$Target.ID),c("Target.ID", "Target.Ligand.ID","Target.Ligand.Ensembl.Gene.ID","Ligand.ID")]) # 注意target id为NA的行被删除
target.id.NA #这是啥？？？
table(targets.gene.unique$Ensembl %in% unique(target.id.NA$Target.Ligand.Ensembl.Gene.ID))
table(ligands.id.uniqe$Ligand.id %in% unique(target.id.NA$Target.Ligand.ID))
table(unique(target.id.NA$Target.Ligand.ID) %in% ligands.id.uniqe$Ligand.id)
# 注意这些ligand并未被记录为target！
all.interactions.targets  <-  all.interactions.targets[!is.na(all.interactions.targets$Target.ID),] 
table(complete.peptide.ligand$Ligand.id %in% ligands.id.uniqe$Ligand.id)
table(all.interactions.targets$Ligand.ID %in% ligands.id.uniqe$Ligand.id) # 比complete.peptide.ligand数据库记录ligands的要多！
```

比较targets数目，基于targets ids

```{r fig.width=6,fig.height=6}
# 发现interactions里的targets id 与 target.and.family里的targets id关系
display_venn(list(interactions.targets.ids.Mouse = na.omit(unique(all.interactions.targets$Target.ID)),
                  targets.ids = na.omit(unique(target.and.family$Target.id))),
             fill = c("#999999", "#E69F00"),cat.col = c("#999999", "#E69F00"), cat.default.pos = "text")
display_venn(list(interactions.targets.ids.Mouse = na.omit(unique(all.interactions.targets$Target.ID)),
                  targets.ids.Mouse = na.omit(unique(targets.gene.unique$Target.ID))),
             fill = c("#999999", "#E69F00"),cat.col = c("#999999", "#E69F00"), cat.default.pos = "text")
# targets.gene.unique 里的targets不能全部包括interactions里targets，这69个interactions.targets.ids.Mouse独有的是啥？是不是ligand为内源peptide的？
unique.7 <- na.omit(unique(all.interactions.targets$Target.ID))[!na.omit(unique(all.interactions.targets$Target.ID)) %in%  na.omit(unique(targets.gene.unique$Target.ID))]
all.interactions.targets[all.interactions.targets$Target.ID %in% unique.7,]
target.and.family[target.and.family$Target.id %in% unique.7,] # 发现可能是target.and.family里targets注释信息缺失造成的，考虑加进去！
```

```{r}
# interactions里独有的这7个targetd id
targets.new.id.unique <- all.interactions.targets[all.interactions.targets$Target.ID %in% unique.7, c( "Target.ID","Target.Ensembl.Gene.ID")]
targets.new.id.unique <- targets.new.id.unique[targets.new.id.unique$Target.Ensembl.Gene.ID != "",] 
targets.new.id.unique <- targets.new.id.unique[!duplicated(targets.new.id.unique$Target.ID),]
dim(targets.new.id.unique) # 剩余6个
targets.new.ensembl.ids <- unlist(lapply(targets.new.id.unique$Target.Ensembl.Gene.ID, function(x){   unlist(strsplit(x,split = "|",fixed = TRUE)) })) # 提取出所有ensembl ids
table(targets.new.ensembl.ids %in% targets.gene.unique$Ensembl) # 基本上targets.new里的基因都被包含了，所以不再去加到targets中！不包含的，可能是因为新旧ID的问题。
targets.new.ensembl.ids[!targets.new.ensembl.ids %in% targets.gene.unique$Ensembl] # 多出来3个
targets.new.ensembl.ids[!targets.new.ensembl.ids %in% gtf$Ensembl] #都存在于gtf中
targets.new.id.unique$Target.Ensembl.Gene.ID <- stringr::str_replace_all(targets.new.id.unique$Target.Ensembl.Gene.ID, "\\|", ";")
dim(targets.new.id.unique)
colnames(targets.new.id.unique) <- c("Target.id", "Mouse.Ensembl.Gene")

targets.new.id.unique$Type <- "Unknown"
targets.new.id.unique$Family.name <- "Unknown"
targets.new.id.unique$MGI <- update_IDs(old = targets.new.id.unique$Mouse.Ensembl.Gene, db = gtf, from = "Ensembl", to = "MGI", split = ";", fixed = TRUE)
targets.new.id.unique$Symbol <- update_IDs(old = targets.new.id.unique$Mouse.Ensembl.Gene, db = gtf, from = "Ensembl", to = "Symbol", split = ";", fixed = TRUE)
targets.new.id.unique$Entrez <- update_IDs(old = targets.new.id.unique$Mouse.Ensembl.Gene, db = gtf, from = "Ensembl", to = "EntrezID", split = ";", fixed = TRUE)

targets.gene.unique %>%
  group_by(Target.ID) %>%
  mutate(Ensembl = paste0(unique(Ensembl), collapse = ";"), 
         MGI = paste0(unique(MGI), collapse = ";"),
         Symbol = paste0(unique(Symbol), collapse = ";"),
         Entrez = paste0(unique(Entrez), collapse = ";"),
         Type = paste0(unique(Type), collapse = ";"),
         Family = paste0(unique(Family), collapse = ";")) %>%
  filter(row_number()==1) %>%  # 仅保留第一行
  ungroup() -> targets.id.unique

colnames(targets.new.id.unique) <- c("Target.ID", "Ensembl","Type" ,"Family",  "MGI","Symbol","Entrez")
targets.new.id.unique <- targets.new.id.unique[,colnames(targets.id.unique)]
targets.new.id.unique <- targets.new.id.unique[!targets.new.id.unique$Ensembl %in% targets.id.unique$Ensembl,] # 去除重复的Ensembl
targets.full.id.unique <- rbind(targets.id.unique, targets.new.id.unique)
dim(targets.full.id.unique) # 合并了target文件与interactions文件里的targets
table(duplicated(targets.full.id.unique$Ensembl))
```

哪些基因既是targets，又是ligands

```{r}
intersect(targets.gene.unique$Symbol, ligands.gene.uniqe$Symbol)
```

比较ligands数目，基于ligand ids

```{r fig.width=4,fig.height=4}
# 发现interactions里的targets id 与 target.and.family里的targets id关系
display_venn(list(interactions.ligand.ids.Mouse = na.omit(unique(all.interactions.targets$Ligand.ID)),
                  ligand.ids = na.omit(unique(ligand.ID.mapping$Ligand.id))),
             fill = c("#999999", "#E69F00"),cat.col = c("#999999", "#E69F00"))
display_venn(list(interactions.ligand.ids.Mouse = na.omit(unique(all.interactions.targets$Ligand.ID)),
                  ligand.ids.Mouse = na.omit(unique(ligands.id.uniqe$Ligand.id))),
             fill = c("#999999", "#E69F00"),cat.col = c("#999999", "#E69F00"))
# ligands.id.uniqe 里的ligand不能全部包括interactions里ligands，这x个应该是非peptide ligands
```


```{r}
interactions <- all.interactions.targets[,c("Target.ID","Target","Target.Gene.Symbol","Target.UniProt.ID","Target.Ensembl.Gene.ID","Ligand", "Ligand.ID", "Approved", "Type", "Action", "Primary.Target" )]
dim(interactions)
interactions <- dplyr::distinct(interactions, .keep_all = TRUE) # 提取唯一行
# 依据 target id和 ligand id 过滤数据
interactions <- interactions[interactions$Target.ID %in% targets.full.id.unique$Target.ID | interactions$Ligand.ID %in% ligands.id.uniqe$Ligand.id,]
dim(interactions)
table(interactions$Type, interactions$Action)

# 重复的targets id
interactions$LT_Pair_ID <- paste0(interactions$Ligand.ID, "_",interactions$Target.ID)
interactions.duplicated.pairs <- janitor::get_dupes(interactions, LT_Pair_ID) # 列出所有重复行
# 发现重复的行，两者的区别在于Type和Action或 Primary.Target 有区别，别的都一样！
dim(interactions.duplicated.pairs)
# 这个问题不知道如何解决了！！！
# 决定，去除Action和Primary.Target两列
interactions <- interactions[,c("Target.ID","Target","Target.Gene.Symbol","Target.UniProt.ID","Target.Ensembl.Gene.ID","Ligand", "Ligand.ID", "Approved", "Type", "LT_Pair_ID")]
dim(interactions)
interactions <- dplyr::distinct(interactions, .keep_all = TRUE) # 提取唯一行
dim(interactions)
janitor::get_dupes(interactions, LT_Pair_ID)
# 还是有重复的 LT_Pair_ID，合并Type值
interactions <- interactions %>%
  group_by(LT_Pair_ID) %>%
  mutate(Type = paste0(Type, collapse = ";")) %>%
  ungroup()
interactions <- dplyr::distinct(interactions, .keep_all = TRUE) 
table(interactions$Type)
table(interactions$Target.ID %in% targets.full.id.unique$Target.ID) # 全部
table(interactions$Ligand.ID %in% ligands.id.uniqe$Ligand.id) # 只有44个，1056个不在
interactions[!interactions$Target.ID %in% targets.full.id.unique$Target.ID,]
```


```{r}
# 修正interactions 里的target 和 Ligand ID的基因信息
interactions$Target.Ensembl <- targets.full.id.unique$Ensembl[match(interactions$Target.ID, targets.full.id.unique$Target.ID)]
interactions$Target.Type <- targets.full.id.unique$Type[match(interactions$Target.ID, targets.full.id.unique$Target.ID)]
interactions$Target.Family <- targets.full.id.unique$Family[match(interactions$Target.ID, targets.full.id.unique$Target.ID)]

interactions$Ligand.Ensembl <- ligands.id.uniqe$Ensembl.ID[match(interactions$Ligand.ID,ligands.id.uniqe$Ligand.id)]
interactions <- interactions[,c("LT_Pair_ID", "Target.ID","Target","Target.Ensembl","Target.Type", "Target.Family", "Ligand.ID", "Ligand", "Ligand.Ensembl","Approved", "Type")]
interactions$Target.Symbol <- update_IDs(old = interactions$Target.Ensembl, db = gtf, from = "Ensembl", to = "Symbol", split = ";", fixed = TRUE)
interactions$Ligand.Symbol <- update_IDs(old = interactions$Ligand.Ensembl, db = gtf, from = "Ensembl", to = "Symbol", split = ";", fixed = TRUE)
interactions <- interactions[,c("LT_Pair_ID","Target.ID","Target","Target.Ensembl", "Target.Symbol","Target.Type", "Target.Family", "Ligand.ID", "Ligand", "Ligand.Ensembl","Ligand.Symbol","Approved", "Type")]
```



```{r}
table(duplicated(interactions$Target.ID))
Mouse_GuideToPHARMACOLOGY_Interactions_LigandTargetsFull <- interactions
usethis::use_data(Mouse_GuideToPHARMACOLOGY_Interactions_LigandTargetsFull, overwrite = TRUE)
```
注意，Mouse_GuideToPHARMACOLOGY_Interactions-Ligand-Targets-Full.rda收录的Ligand和Targets要多于Mouse_GuideToPHARMACOLOGY_LigandsPeptides和Mouse_GuideToPHARMACOLOGY_Targets的！

```{r}
# 拆Ensembl ID
interactions.list <- split(interactions, f = interactions$LT_Pair_ID)
interactions.list <- lapply(interactions.list, function(x){
  ensembl.id <- unlist(strsplit(x$Target.Ensembl, split = ";"))
  x <- x[rep(1,length(ensembl.id)),]
  x$Target.Ensembl <- ensembl.id
  if(!is.na(x$Ligand.Ensembl[1]) & grepl(";",x$Ligand.Ensembl[1])){
     ligand.ensembl.id <-  unlist(strsplit(x$Ligand.Ensembl[1], split = ";"))
     x <- x[rep(1:length(ensembl.id),length(ligand.ensembl.id)),]
     x$Ligand.Ensembl <- rep(ligand.ensembl.id, each =length(ensembl.id))
  }
  return(x)
})
interactions.gene <- Reduce(rbind,interactions.list)
dim(interactions.gene)
```

```{r}
interactions.gene$Target.Symbol <- gtf$Symbol[match(interactions.gene$Target.Ensembl, gtf$Ensembl)]
interactions.gene$Ligand.Symbol <- gtf$Symbol[match(interactions.gene$Ligand.Ensembl, gtf$Ensembl)]
head(sort(table(interactions.gene$Target.Symbol),decreasing = TRUE),30)
head(sort(table(interactions.gene$Ligand.Symbol),decreasing = TRUE),30)
```
```{r}
Mouse_GuideToPHARMACOLOGY_Interactions_LigandTargetsFull_Expanded <- interactions.gene
usethis::use_data(Mouse_GuideToPHARMACOLOGY_Interactions_LigandTargetsFull_Expanded, overwrite = TRUE)
```

```{r}
# interactions.gene.peptides <- dplyr::filter(interactions.gene, grepl("^ENSMUSG",Target.Ensembl),grepl("^ENSMUSG",Ligand.Ensembl))
# dim(interactions.gene.peptides)
# interactions.gene.peptides$LT_Pair_Symbol <- paste0(interactions.gene.peptides$Ligand.Symbol, "_",interactions.gene.peptides$Target.Symbol)
# # janitor::get_dupes(interactions.gene.peptides,LT_Pair_Symbol)
# interactions.gene.peptides <- interactions.gene.peptides[!duplicated(interactions.gene.peptides$LT_Pair_Symbol),] # 默认保留重复值的第一个
# interactions.gene.peptides$Target.Type <- targets.id.unique$Type[match(interactions.gene.peptides$Target.ID, targets.id.unique$Target.id)] # 修正Target.Type
# interactions.gene.peptides$Target.Family <- targets.id.unique$Family.name[match(interactions.gene.peptides$Target.ID, targets.id.unique$Target.id)]
# dim(interactions.gene.peptides)
# table(interactions.gene.peptides$Target.Type)
# head(sort(table(interactions.gene.peptides$Target.Symbol),decreasing = TRUE),30)
# head(sort(table(interactions.gene.peptides$Ligand.Symbol),decreasing = TRUE),30)
# intersect(interactions.gene.peptides$Target.Symbol, interactions.gene.peptides$Ligand.Symbol)
```

**基于对应的targets相似性，对ligand进行分群**

```{r}
LT.mat <- as.data.frame.matrix(table(interactions.gene$Ligand.Symbol, interactions.gene$Target.Symbol)) # ligand - target matrix
# table(LT.mat[1,],LT.mat[2,])/ncol(LT.mat)
res <- matrix(data = 0, nrow = nrow(LT.mat), ncol = nrow(LT.mat), dimnames = list(rownames(LT.mat), rownames(LT.mat)))
system.time(for (i in 1:nrow(res)) {
  for (j in i:nrow(res)) {
    data.sub <- unlist(LT.mat[i,],use.names = FALSE) + unlist(LT.mat[j,],use.names = FALSE)
    res[i,j] <- sum(data.sub == 2)
  }
})

library(pheatmap)
for (i in 1:nrow(res)) {
  for (j in 1:i) {
    res[i,j] <- res[j,i]
  }
}
remove.iterms.index <- unname(apply(res != 0, 2, sum) != 1)
res <- res[remove.iterms.index, remove.iterms.index]
pdf(file = "./Mouse_Ligands-Heatmap-By-Shared-Targets.pdf",width = 10,height = 10)
pheatmap::pheatmap(res)
dev.off()
```

**基于对应的ligands相似性，对target进行分群**

```{r}
LT.mat <- t(LT.mat)
# table(LT.mat[1,],LT.mat[2,])/ncol(LT.mat)
res <- matrix(data = 0, nrow = nrow(LT.mat), ncol = nrow(LT.mat), dimnames = list(rownames(LT.mat), rownames(LT.mat)))
system.time(for (i in 1:nrow(res)) {
  for (j in i:nrow(res)) {
    data.sub <- unlist(LT.mat[i,],use.names = FALSE) + unlist(LT.mat[j,],use.names = FALSE)
    res[i,j] <- sum(data.sub == 2)
  }
})

library(pheatmap)
for (i in 1:nrow(res)) {
  for (j in 1:i) {
    res[i,j] <- res[j,i]
  }
}
remove.iterms.index <- unname(apply(res != 0, 2, sum) != 1)
res <- res[remove.iterms.index, remove.iterms.index]
pdf(file = "./Mouse_Targets-Heatmap-By-Shared-Ligands.pdf",width = 40,height = 40)
pheatmap::pheatmap(res)
dev.off()
```

**保存数据**

```{r}
# targets.full.id.unique 里的Ensembl ID有些多有个
targets.full.id.unique %>%
  group_by(Target.ID) %>%
  dplyr::slice(rep(1:n(), each = stringr::str_count(Ensembl,"ENSMUSG"))) %>%
  mutate(Ensembl = trimws(unlist(strsplit(Ensembl[1], split = ";", fixed = TRUE))))  %>%
  ungroup() -> targets.full.id.unique.splited
  
targets.full.gene.unique <- targets.full.id.unique.splited %>%
  group_by(Ensembl) %>%
  mutate(Target.ID = paste0(unique(Target.ID), collapse = ";"),
         Type  = paste0(unique(Type), collapse = ";"),
         Family = paste0(unique(Family), collapse = ";"),
         MGI = paste0(unique(MGI), collapse = ";")) %>%
  filter(row_number()==1) %>%  # 仅保留第一行
  ungroup()
targets.full.gene.unique$MGI <- gtf$Symbol[match(targets.full.gene.unique$Ensembl, gtf$Ensembl)]
```

```{r}
Mouse_GuideToPHARMACOLOGY_Interactions_OnlyTargetsFull_Expanded <- targets.full.gene.unique
usethis::use_data(Mouse_GuideToPHARMACOLOGY_Interactions_OnlyTargetsFull_Expanded, overwrite = TRUE)
```

带full，表示收录的基因数目会更多一些！


## UniProt

similar Protein family

```{r}
load("../data/UniProt_ProteinFamily_AllSpecies.rda")
res <- UniProt_ProteinFamily_AllSpecies
protein.mouse <- res[unlist(lapply(res,function(x)any(x$Species == "MOUSE")))]
protein.mouse <- lapply(protein.mouse, function(x)x[x$Species == "MOUSE",])
protein.mouse <- Reduce(rbind, protein.mouse)
```

```{r}
protein.mouse$Species <- NULL
group_by(protein.mouse, Gene) %>%
  mutate(family = paste0(family, collapse = ";")) %>%
  filter(row_number() == 1) %>%  # 仅保留第一行
  ungroup() -> protein.mouse
```


```{r}
protein.mouse$Ensembl.Added <- gtf$Ensembl[match(protein.mouse$Protein, gtf$UniProt)]
protein.mouse$Symbol.Added <- gtf$Symbol[match(protein.mouse$Protein, gtf$UniProt)]

UniProt_ProteinFamily_Mouse <- protein.mouse
usethis::use_data(UniProt_ProteinFamily_Mouse, overwrite = TRUE)
```


## VerSeDa: vertebrate secretome database

预测的结果，很多应该都是不靠谱的！

2025.04.16 跳过了这部分！

http://genomics.cicbiogune.es/VerSeDa/FullOrganisms.php

```{r}
VerSeDa <- read.delim("../inst/extdata/VerSeDa-vertebrate-secretome-database/mus-musculus-GRCm38/resultsTable.txt.gz", stringsAsFactors = FALSE)
score.list <- strsplit(VerSeDa$WoLF.PSORT, split = ",")
extr.score <- unlist(lapply(score.list, function(x){
  if (any(grepl("extr ",x))) {
    extr.score <- x[grepl("extr ",x)]
    if (length(extr.score) != 1) {
      print(x)
      stop("more than 1 extr found!")
    }
    return(as.numeric(trimws(gsub("extr ","",extr.score))))
  }else{
    return(NA)
  }
}))

VerSeDa$WoLF.PSORT.extr.score <- extr.score
gene.level <- split(VerSeDa, f = VerSeDa$Gene)
gene.df <- lapply(gene.level, function(x){
  df <- as.data.frame(apply(x,2,function(x)paste0(x,collapse = ";")),stringsAsFactors = FALSE)
  colnames(df)[1] <- x$Gene[1]
  df <- as.data.frame(t(df),stringsAsFactors = FALSE)
  df$Gene <- x$Gene[1]
  return(df)
})
gene.con <- Reduce(rbind, gene.df)
rownames(gene.con) <- NULL
colnames(gene.con)[2] <- "Ensembl"
gene.con <- gene.con[,c("Ensembl", "Uniprot","SignalP","TargetP","TMHMM","PredGPI","SecretomeP","WoLF.PSORT.extr.score")]

VerSeDa_Mouse_SecretomePrediction <- gene.con
usethis::use_data(VerSeDa_Mouse_SecretomePrediction, overwrite = TRUE)
```


## CellChat V2

github: https://github.com/jinworks/CellChat

url: https://github.com/jinworks/CellChat/tree/main/data

```{r}
load("../inst/extdata/CellChatV2/CellChatDB.mouse.rda")
interactions <- CellChatDB.mouse$interaction
rownames(interactions) <- NULL
```

```{r}
# first remove rows in interactions which the ligands symbol not existed in gtf
ligands <- unique(trimws(unlist(strsplit(interactions$ligand.symbol, split = ", "))))
table(ligands %in% gtf$Symbol)
ligands.lost <- ligands[!ligands %in% gtf$Symbol]
# list choose ligands.symbol which not existed in gtf$symbol
interactions$ligand.symbol[unlist(lapply(strsplit(interactions$ligand.symbol, split = ", "), function(x)any(x %in% ligands.lost)))]
ligands.mapping <- c(Cd99 = "Cd99l2", # 可能是~ 
                     # Ccl21 unknown
                     # H2-Q8
                     "H2-Bl" = "H2-T13",
                     # H2-T18
                     # H2-Q9
                     "H2-T26" = "H2-T5",
                     # H2-L
                     Gm10499 = "H2-T7",
                     "H2-T-ps"= "H2-T11-ps",
                     "H2-T27" = "H2-T15",
                     "H2-T25" = "H2-T9"
                     # H2-D
                     # "Raet1a"
                     # "H60a"
                     # Raet1c
                     # "Raet1b"
                     )
ligands.mapping <- ligands.mapping[unname(ligands.mapping) %in% gtf$Symbol]
for (i in names(ligands.mapping)) {
  interactions$ligand.symbol[interactions$ligand.symbol == i] <- ligands.mapping[i]
}
interactions <- interactions[unlist(lapply(strsplit(interactions$ligand.symbol, split = ", "), function(x)all(x %in% gtf$Symbol))),]

# seconda remove rows in interactions which the receptors symbol not existed in gtf
interactions$receptor.symbol <- gsub("_", ", ", interactions$receptor.symbol)
receptors <- unique(trimws(unlist(strsplit(interactions$receptor.symbol, split = ", "))))
table(receptors %in% gtf$Symbol)
receptors.lost <- receptors[!receptors %in% gtf$Symbol]
interactions$receptor.symbol[unlist(lapply(strsplit(interactions$receptor.symbol, split = ", "), function(x)any(x %in% receptors.lost)))]

receptors.mapping <- c(Cd99 = "Cd99l2") # 可能是~ 
receptors.mapping <- receptors.mapping[unname(receptors.mapping) %in% gtf$Symbol]
for (i in names(receptors.mapping)) {
  interactions$receptor.symbol[interactions$receptor.symbol == i] <- receptors.mapping[i]
}
interactions <- interactions[unlist(lapply(strsplit(interactions$receptor.symbol, split = ", "), function(x)all(x %in% gtf$Symbol))),]
```


```{r}
interactions$receptor.Ensembl <- update_IDs(old = interactions$receptor.symbol, db = gtf, from = "Symbol", to = "Ensembl", split = ",", fixed = TRUE)
interactions$ligand.Ensembl <- update_IDs(old = interactions$ligand.symbol, db = gtf, from = "Symbol", to = "Ensembl",  split = "_", fixed = TRUE)

CellChat_Mouse_LigandReceptorsInteractions <- interactions
usethis::use_data(CellChat_Mouse_LigandReceptorsInteractions, overwrite = TRUE)
```

```{r}
lrs <- interactions[,c("ligand.Ensembl", "receptor.Ensembl", "pathway_name", "interaction_name_2")]
colnames(lrs)[4] <- "interaction"

receptor.sub <- lrs[,c(2,3,4)]
table(duplicated(receptor.sub$receptor.Ensembl))

receptor.sub <- Reduce(rbind, lapply(split(receptor.sub, f = receptor.sub$receptor.Ensembl),function(x){
  x$pathway_name <- paste0(unique(x$pathway_name),collapse = ";")
  x$interaction <- paste0(unique(x$interaction),collapse = ";")
  return(x[1,])
}))

receptor.sub <- Reduce(rbind, lapply(split(receptor.sub, f = receptor.sub$receptor.Ensembl),function(x){
  receptors <- unlist(strsplit(x$receptor.Ensembl,split = ";"))
  x <- x[rep(1,length(receptors)),]
  x$receptor.Ensembl <- receptors
  return(x)
}))

receptor.sub <- receptor.sub[!duplicated(receptor.sub),]

receptor.sub <- Reduce(rbind, lapply(split(receptor.sub, f = receptor.sub$receptor.Ensembl),function(x){
  x$pathway_name <- paste0(unique(x$pathway_name),collapse = ";")
  x$interaction <- paste0(unique(x$interaction),collapse = ";")
  return(x[1,])
}))

CellChat_Mouse_Receptors <- receptor.sub
usethis::use_data(CellChat_Mouse_Receptors, overwrite = TRUE)


ligand.sub <- lrs[,c(1,3,4)]
table(duplicated(ligand.sub$ligand.Ensembl))

ligand.sub <- Reduce(rbind, lapply(split(ligand.sub, f = ligand.sub$ligand.Ensembl),function(x){
  x$pathway_name <- paste0(unique(x$pathway_name),collapse = ";")
  x$interaction <- paste0(unique(x$interaction),collapse = ";")
  return(x[1,])
}))

ligand.sub <- Reduce(rbind, lapply(split(ligand.sub, f = ligand.sub$ligand.Ensembl),function(x){
  receptors <- unlist(strsplit(x$ligand.Ensembl,split = ";"))
  x <- x[rep(1,length(receptors)),]
  x$ligand.Ensembl <- receptors
  return(x)
}))

ligand.sub <- ligand.sub[!duplicated(ligand.sub),]

ligand.sub <- Reduce(rbind, lapply(split(ligand.sub, f = ligand.sub$ligand.Ensembl),function(x){
  x$pathway_name <- paste0(unique(x$pathway_name),collapse = ";")
  x$interaction <- paste0(unique(x$interaction),collapse = ";")
  return(x[1,])
}))

CellChat_Mouse_Ligands <- ligand.sub
usethis::use_data(CellChat_Mouse_Ligands, overwrite = TRUE)
```


## CellTalkDB

url: http://tcm.zju.edu.cn/celltalkdb/browser.php

```{r}
lrs <- read.delim("../inst/extdata/CellTalkDB/mouse_lr_pair.txt.gz", stringsAsFactors = FALSE)
table(lrs$ligand_ensembl_gene_id %in% gtf$Ensembl)
lrs[!lrs$ligand_ensembl_gene_id %in% gtf$Ensembl,]
lrs$ligand_ensembl_gene_id[lrs$ligand_ensembl_gene_id == "ENSMUSG00000096873"] <- "ENSMUSG00000073878"

table(lrs$receptor_ensembl_gene_id %in% gtf$Ensembl)
lrs[!lrs$receptor_ensembl_gene_id %in% gtf$Ensembl,]
lrs$receptor_ensembl_gene_id[lrs$receptor_ensembl_gene_id == "ENSMUSG00000048191"] <- "ENSMUSG00000118661"

lrs$ligand_gene_symbol <- gtf$Symbol[match(lrs$ligand_ensembl_gene_id, gtf$Ensembl)]
lrs$receptor_gene_symbol <- gtf$Symbol[match(lrs$receptor_ensembl_gene_id, gtf$Ensembl)]

lrs$ligand_gene_id <- gtf$EntrezID[match(lrs$ligand_ensembl_gene_id, gtf$Ensembl)]
lrs$receptor_gene_id <- gtf$EntrezID[match(lrs$receptor_ensembl_gene_id, gtf$Ensembl)]

CellTalkDB_Mouse_Interactions <- lrs
usethis::use_data(CellTalkDB_Mouse_Interactions, overwrite = TRUE)
```

## NCBI: Gene summary

> https://www.ncbi.nlm.nih.gov/books/NBK3840/table/genefaq.Tc/

[Mus_musculus.ags.gz  ](https://ftp.ncbi.nih.gov/gene/DATA/ASN_BINARY/Mammalia/):	Gene records

**首先将ASN文件转为XML文件**

参考：[9.10 Parsing huge Entrez XML files](https://biopython-cn.readthedocs.io/zh-cn/latest/en/chr09.html)

**install ncbi-tools-bin, and trans ags to xml file:**

```{bash eval=FALSE}
# sudo apt-get install ncbi-tools-bin
gunzip Mus_musculus.ags.gz
gene2xml -b T  -i Mus_musculus.ags -o Mus_musculus.xml
```

**提取基因信息**

```{bash eval=FALSE}
cat Mus_musculus.xml | grep "<Gene-track_geneid>\|<Gene-ref_locus>\|<Entrezgene_summary>" > subset.txt
# 去掉行前的空格
sed  -i 's/^[ ]*//g' subset.txt
```

然后使用R语言提取基因信息，具体流程请见："inst/extdata/NCBI/Human_Gene_Summary/extract-gene-summary.md"文件。

```{r }
load("../data/GENCODE_Mouse_Genes.rda")
```

```{r}
ncbi <- readRDS("../inst/extdata/NCBI/Mouse_Gene_Summary/NCBI_gene_summary_from_ASN_filtered.rds")
table(duplicated(ncbi$Symbol))
table(duplicated(ncbi$gene_id))
table(GENCODE_Mouse_Genes$EntrezID %in% ncbi$EntrezID)
table(GENCODE_Mouse_Genes$Symbol %in% ncbi$Symbol)
table(GENCODE_Mouse_Genes$Ensembl %in% ncbi$Ensembl)
table(GENCODE_Mouse_Genes$Ensembl %in% ncbi$Ensembl.By.OrgHSegdb)
table(GENCODE_Mouse_Genes$MGI %in% gsub("^MGI:", "",ncbi$MGI))
```
```{r}
NCBI_Mouse_GeneSummary <- ncbi
usethis::use_data(NCBI_Mouse_GeneSummary, overwrite = TRUE)
```

## Uniprot: Gene Function
 
下载蛋白的注释信息，具体参考："inst/extdata/UniProt/GeneFunction/Human/提取-UniProt-Protein-summary.md"

```{r}
uniprot <- read.delim("../inst/extdata/UniProt/GeneFunction/Mouse/uniprotkb_mouse_AND_reviewed_true_AND_m_2025_04_16.tsv.gz",stringsAsFactors = FALSE)
uniprot <- uniprot[uniprot$Reviewed == "reviewed",]
uniprot <- uniprot[uniprot$Function..CC. != "",]
uniprot <- uniprot[uniprot$Gene.Names..primary. != "",]
dim(uniprot)

# primary gene name 有多个的，展开后再合并
uniprot.sub <- uniprot[grepl(";",uniprot$Gene.Names..primary.),]
uniprot <- uniprot[!grepl(";",uniprot$Gene.Names..primary.),]

dup.numbers <- unname(sapply(uniprot.sub$Gene.Names..primary.,function(x)length(unlist(strsplit(x,split = ";")))))
uniprot.sub.expand <- uniprot.sub[rep(1:nrow(uniprot.sub),dup.numbers),]
uniprot.sub.expand$Gene.Names..primary. <- trimws(unlist(strsplit(uniprot.sub$Gene.Names..primary.,split = ";")))
uniprot <- rbind(uniprot, uniprot.sub.expand)
```


```{r}
uniprot$Reviewed <- NULL
uniprot$Organism <- NULL
uniprot$Gene.Names <- NULL
uniprot$Gene.encoded.by <- NULL
colnames(uniprot) <- c("Entry", "Entry.Name", "Protein.names", "FunctionCC" ,"Primary.Gene.Names")
table(uniprot$Primary.Gene.Names %in% GENCODE_Human_Genes$Symbol)
table(duplicated(uniprot$Primary.Gene.Names))
```
问题是，有少量基因是重复的！

```{r}
uniprot$Ensembl.Added <- GENCODE_Mouse_Genes$Ensembl[match(uniprot$Primary.Gene.Names, GENCODE_Mouse_Genes$Symbol)]
UniProt_Mouse_GeneFunction <- uniprot
usethis::use_data(UniProt_Mouse_GeneFunction, overwrite = TRUE)
```


## GENEONTOLOGY

> [详解如何获取物种所有基因对应的GO注释](https://cloud.tencent.com/developer/article/1625243)

下载位置： https://current.geneontology.org/products/pages/downloads.html

https://geneontology.org/docs/download-ontology/ 页面下载： https://purl.obolibrary.org/obo/go/go-basic.obo

```{r}
df <- read.delim("../inst/extdata/GENEONTOLOGY/mgi.gaf.gz",skip = 36,header = FALSE)
# https://geneontology.org/docs/go-annotation-file-gaf-format-2.2/
colnames(df) <- c("DB", "ProteinID", "Symbol","Qualifier","GOID","Reference", "EvidenceCode","WithOrFrom","Aspect", 
                  "GeneProducts","Synonym","Type","Taxon", "Date","AssignedBy","AnnotationExtension","GeneProductsFromID")
id.unique <- unique(df$ProteinID)
# write.csv(id.unique, file = "IDs.unique.csv",row.names = FALSE,quote = FALSE)
# 先用proteinID做第一次id更新
id.res <-read.delim("./MGIBatchReport_IDs.txt",stringsAsFactors = FALSE)
id.lost <- id.res[id.res$Input.Type == "",]
id.lost$Input <- gsub("-\\d+","",id.lost$Input)
# write.csv(id.lost$Input, file = "IDs.2.unique.csv",row.names = FALSE,quote = FALSE)
# 用没有匹配上的，去除末尾的-数字后缀，在做一个id更新
id.2.res <-read.delim("./MGIBatchReport_IDs-2.txt",stringsAsFactors = FALSE)
table(id.2.res$Input.Type)
id.mapped <- rbind(id.res[id.res$Input.Type != "",], id.2.res[id.2.res$Input.Type != "",])
# 最后的id匹配结果
id.mapped <- id.mapped[!duplicated(id.mapped$Input),]
table(grepl("^MGI",id.mapped$MGI.Gene.Marker.ID))
#janitor::get_dupes(id.mapped,  MGI.Gene.Marker.ID)  
#janitor::get_dupes(id.mapped,  Symbol)
#经查，每个MGI id和每个Symbol是一一对应的。
id.mapped.unique <- distinct(id.mapped, MGI.Gene.Marker.ID, Symbol)
table(duplicated(id.mapped.unique$MGI.Gene.Marker.ID))
table(duplicated(id.mapped.unique$Symbol))
id.sym.mapping <- id.mapped.unique$Symbol
names(id.sym.mapping) <- id.mapped.unique$MGI.Gene.Marker.ID

df$ProteinID <- gsub("-\\d+","",df$ProteinID)
df <- df[df$ProteinID %in% id.mapped$Input,]
df$ProteinID <- id.mapped$MGI.Gene.Marker.ID[match(df$ProteinID, id.mapped$Input)]
df$Symbol <- id.mapped$Symbol[match(df$ProteinID, id.mapped$Input)]

df <- distinct(df, ProteinID, GOID, .keep_all = TRUE) # 去除重复的
dim(df)

# 测试了symbol，结论是用ProteinID列就可以了！
# symbol.unique <- unique(df$Symbol)
# write.csv(symbol.unique, file = "Symbols.unique.csv",row.names = FALSE,quote = FALSE)
# sys.res <- read.delim("./MGIBatchReport_Symbols.txt",stringsAsFactors = FALSE)
# id.lost.symbols <- unique(df$ProteinID[df$ProteinID %in% id.lost$Input])
# sys.res.lost.lost <- sys.res[sys.res$Input %in% id.lost.symbols,]
# table(sys.res.lost.lost$Input.Type)

library(ontologyIndex)
gs <- get_OBO("../inst/extdata/GENEONTOLOGY/go-basic.obo")
table(df$GOID %in% names(gs$name))
df$GOName <- gs$name[df$GOID]
```

```{r}
# 为什么不用sumbol，因为有30多个NA值
df.sub <- df[,c("ProteinID","GOID","GOName","Aspect")]
# remove 3 root: GO:0005575:cellular_component, GO:0003674:molecular_function, GO:0008150: biological_process
df.sub <- df.sub[!df.sub$GOID %in% c("GO:0005575", "GO:0003674", "GO:0008150"),]
res <- list()
aspect.name <- c("MolecularFunction","BiologicalProcess","CellularComponent")
names(aspect.name) <- c("F","P","C")
for (i in unique(df.sub$Aspect)) {
  df.sub.sub <- df.sub[df.sub$Aspect == i,]
  df.sub.sub.aggr <- Aggregate_df(df = df.sub.sub, id = "ProteinID")
  df.sub.sub.aggr$Aspect <- NULL
  colnames(df.sub.sub.aggr)[2:3] <- paste(aspect.name[i],colnames(df.sub.sub.aggr)[2:3],sep = "_")
  res[[aspect.name[i]]] <- df.sub.sub.aggr
}
```


```{r}
# saveRDS(res,file = "temp.rds")
# run codes bellow under a server!
# results <- merge(res$MolecularFunction, res$CellularComponent, all= TRUE, by = "Symbol")
# results <- merge(results, res$BiologicalProcess, all= TRUE, by = "Symbol")
# results[is.na(results)] <- "-"
```

```{r}
GENEONTOLOGY_Mouse <- readRDS("./mouse.go.rds")
GENEONTOLOGY_Mouse$Symbol <- id.sym.mapping[GENEONTOLOGY_Mouse$ProteinID]
colnames(GENEONTOLOGY_Mouse)[1] <- "MGI"
GENEONTOLOGY_Mouse <- GENEONTOLOGY_Mouse[, c( "MGI", "Symbol", "MolecularFunction_GOID",  "MolecularFunction_GOName", "CellularComponent_GOID", "CellularComponent_GOName", "BiologicalProcess_GOID" , "BiologicalProcess_GOName")]

load("../data/GENCODE_Mouse_Genes.rda")
table(GENEONTOLOGY_Mouse$Symbol %in% GENCODE_Mouse_Genes$Symbol)
GENEONTOLOGY_Mouse$Ensembl.Added <- GENCODE_Mouse_Genes$Ensembl[match(GENEONTOLOGY_Mouse$Symbol, GENCODE_Mouse_Genes$Symbol)]
usethis::use_data(GENEONTOLOGY_Mouse, overwrite = TRUE)
```


